### PR TITLE
SC-18 # Allow customers to configure CORS headers

### PR DESCRIPTION
**Pull request by mymattcarroll**
### Added
- SC-24: Display of CORS for `bm server info` cmd
- SC-24: CORS to `bm server serverless` when generating a serverless project
- SC-24: CORS to `bm server serve` when locally hosting routes
  - Will still run if CORS is not setup, and will use [hapi](http://hapijs.com/api#route-options) defaults
### Code Review Notes
- Probably worth reviewing these one commit at a time
- Known incomplete work to be tackled in future:
  - [ ] Monitor this [serverless issue](https://github.com/serverless/serverless/issues/2476) and ensure the `bm server serverless` cmd is compatible with the fix

### CODE REVIEW

File: test/cors/read.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      return Promise.resolve({
+        cors: CORS
+      })
+    })
+  })
+
+  return read(CWD)
+})
+
+test('Should handle an unitinitalised config file', (t) => {
```
Comment (jokeyrhyme) [inline test/cors/read.js]: "unitinitalised" typo?

File: lib/cors/validate.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import type {CorsConfiguration} from './read.js'
+*/
+
+const HELP = ', see documentation for information on how to configure cors.'
+
+function validateCors (
+  cors /* : CorsConfiguration */
+) /* : Promise<CorsConfiguration> */ {
+  if (!cors) {
+    return Promise.reject(new Error('Must specify cors configuration' + HELP))
```
Comment (jokeyrhyme) [inline lib/cors/validate.js:18]: It should be okay to not have a CORS configuration. After all, that's what machine-to-machine or Lambda-inception use cases, or even just the common case of native-only front-ends will be likely to do.

File: lib/cors/validate.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  }
+  if (!cors.origins || !Array.isArray(cors.origins) || !cors.origins.length) {
+    return Promise.reject(new Error('Must specify at least a single allowable origin in cors configuration' + HELP))
```
Comment (jokeyrhyme) [inline lib/cors/validate.js:22]: You should be allowed to have either 0 origins, or 1 origin, due to the limitations we've discovered with AWS Lambda and/or Serverless.

Once that's been sorted out, we can allow multiple origins.

Comment (jokeyrhyme) [reply lib/cors/validate.js:22]: So, they should be a good production origin in, which we use for production deployments. And we automatically just use "*" for the local development server (if they've specified an origin for production) just to regain some symmetry.

File: lib/serve.js [hunk 4]
```diff
@@ -37,9 +37,16 @@ function startServer (options) {
   const server = new Hapi.Server()
   server.connection({ port: options.port })
 
+  let cors = true // development only, not recommended in production
+  if (options.cors && options.cors.origins) {
+    cors = {
+      origin: options.cors.origins,
```
Comment (jokeyrhyme) [inline lib/serve.js]: We have options here, either copy the single origin that they want to use for production, or use the presence of a single origin to decide to use a "*". If they've got no origins, then our local server probably shouldn't either.

**Review by jokeyrhyme — CHANGES_REQUESTED** [review]

File: test/cors/read.js (hunk 1 above)
Comment (mymattcarroll) [reply test/cors/read.js]: dam

**Comment by mymattcarroll** [comment]
@jokeyrhyme, what about adding a new `--no-cors` flag to the `info` (and eventually the `deploy`) cmd to allow you run them without cors configuration set up?

The `serve` and `serverless` cmds already allows for this.

File: lib/serve.js (hunk 4 above)
Comment (mymattcarroll) [inline lib/serve.js]: The `serve` cmd already allows for this.

**Comment by jokeyrhyme** [comment]
@mymattcarroll it will be a very common case to not have CORS configured. I fear hiding this behind a flag would just end up in a similar situation to the `--force` flag that we recently improved.

**Pull request closed without merging**